gh-106670: Allow Pdb to move between chained exceptions

Carreau opened the pull request:
Allow Pdb to move between chained exception.

This lets Pdb receive an exception, instead of a traceback, and when
this is the case and the exception are chained, the new `exceptions` command
allows to both list (no arguments) and move between the chained exceptions.

That is to say if you have something like

    def out():
        try:
            middle()                                # B
        except Exception as e:
            raise ValueError("foo(): bar failed")   # A

    def middle():
        try:
            return inner(0)                         # D
        except Exception as e:
            raise ValueError("Middle fail") from e  # C

    def inner(x):
        1 / x                                       # E

Only A was reachable after calling `out()` and doing post mortem debug.

With this all A-E points are reachable with a combination of up/down,
and ``exceptions <number>``.

This also change the default behavior of ``pdb.pm()``, as well as
`python -m pdb <script.py>` to receive `sys.last_exc` so that chained
exception navigation is enabled.

We do follow the logic of the ``traceback`` module and handle the
``_context__`` and ``__cause__`` in the same way. That is to say, we try
``__cause__`` first, and if not present chain with ``__context__``. In
the same vein, if we encounter an exception that has
``__suppress_context__`` (like when ``raise ... from None``), we do stop
walking the chain.

Some implementation notes:

  - We do handle cycle in exceptions
  - cleanup of references to tracebacks are not cleared in ``forget()``, as
    ``setup()`` and ``forget()`` are both for setting a single
    exception.
  - We do not handle sub-exceptions of exception groups.
  - We ensure we do not hold references to exceptions too long with a new
    context manager.
  - Have the MAX_CHAINED_EXCEPTION_DEPTH class variable to control the
    maximum number we allow


Closes gh-106670

Carreau committed:
Allow Pdb to move between chained exception.

This lets Pdb receive an exception, instead of a traceback, and when
this is the case and the exception are chained, the new `exceptions` command
allows to both list (no arguments) and move between the chained exceptions.

That is to say if you have something like

    def out():
        try:
            middle()                                # B
        except Exception as e:
            raise ValueError("foo(): bar failed")   # A

    def middle():
        try:
            return inner(0)                         # D
        except Exception as e:
            raise ValueError("Middle fail") from e  # C

    def inner(x):
        1 / x                                       # E

Only A was reachable after calling `out()` and doing post mortem debug.

With this all A-E points are reachable with a combination of up/down,
and ``exception <number>``.

This also change the default behavior of ``pdb.pm()``, as well as
`python -m pdb <script.py>` to receive `sys.last_exc` so that chained
exception navigation is enabled.

We do follow the logic of the ``traceback`` module and handle the
``_context__`` and ``__cause__`` in the same way. That is to say, we try
``__cause__`` first, and if not present chain with ``__context__``. In
the same vein, if we encounter an exception that has
``__suppress_context__`` (like when ``raise ... from None``), we do stop
walking the chain.

Some implementation notes:

 - We do handle cycle in exceptions
 - cleanup of references to tracebacks are not cleared in ``forget()``, as
   ``setup()`` and ``forget()`` are both for setting a single
   exception.
 - We do not handle sub-exceptions of exception groups.

Closes gh-106670

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -206,6 +206,8 @@ def namespace(self):
 
 class Pdb(bdb.Bdb, cmd.Cmd):
+    _chained_exceptions = []
+    _chained_index = 0
```
Could you change this variable to `_chained_exception_index`?

Carreau replied:
Done.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1073,6 +1086,26 @@ def _select_frame(self, number):
+    def do_exceptions(self, arg):
+        if not arg or arg == "list":
+            for ix, exc in enumerate(self._chained_exceptions):
+                print(">" if ix == self._chained_index else " ", ix, repr(exc))
```
I belive you should use `self.message()` instead of `print()` in pdb interface, to make redirection possible, which means you'll probably need to format the string, instead of sending them as separate arguments.

Carreau replied:
You are right. Done.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1073,6 +1086,26 @@ def _select_frame(self, number):
+            except ValueError:
+                self.error("Argument must be an integer")
+                return
+                pass
```
This pass is redundant.

Carreau replied:
Yes, not sure why it sneaked through.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1073,6 +1086,26 @@ def _select_frame(self, number):
+                self.error("Argument must be an integer")
+                return
+                pass
+            else:
```
Why do we need the `else` here?

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1073,6 +1086,26 @@ def _select_frame(self, number):
+                return
+                pass
+            else:
+                if 0 <= arg < len(self._chained_exceptions):
```
Instead of a `return` in normal case, I think it's better to just 
```python
if 0 <= arg < len(self._chained_exceptions):
    # do you stuff
else:
    self.error("XX")
```

Carreau replied:
Updated to use your suggestion.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1895,6 +1928,10 @@ def post_mortem(t=None):
     If no traceback is given, it uses the one of the exception that is
     currently being handled (an exception must be being handled if the
     default is to be used).
+
```
`post_mortem` is a public interface so we need to change the comment

> Enter post-mortem debugging of the given *traceback* object.

Carreau replied:
Done.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1912,7 +1949,9 @@ def post_mortem(t=None):
 def pm():
     """Enter post-mortem debugging of the traceback found in sys.last_traceback."""
-    if hasattr(sys, 'last_exc'):
+    if hasattr(sys, "last_value"):
```
`last_value` is actually deprecated. This is a new feature and will only be applicable to python3.13+, so we should keep the logic and send the exception instead of traceback.

Carreau replied:
Ah good point. Sorry I missed this.

gaogaotiantian commented on Misc/NEWS.d/next/Library/2023-07-12-10-59-08.gh-issue-106670.goQ2Sy.rst:
```diff
@@ -0,0 +1,47 @@
+Allow Pdb to move between chained exceptions on post_mortem debugging.
```
I don't think you should write a detailed change in news entry. The news entry should be compact and simple, preferably a sentence or two. I think this actually breaks the docs test.

Carreau replied:
Moved to Doc/library/pdb.rst where I describe the new exceptions command.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1073,6 +1086,26 @@ def _select_frame(self, number):
+                    self._chained_index = arg
+                    self.setup(None, self._chained_exceptions[arg].__traceback__)
+                    self.curindex = 0
+                    self._select_frame(0)
```
Why set the frame to the top instead of where the actual exception happened?

Carreau replied:
Not sure, I thin it was just to print the current frame when doing `exceptions <number>` maybe ? but I now directly print the frame and let `self.setup` select the right frame.

gaogaotiantian reviewed (changes requested):
Left some comments. For this feature to be merged, you'll need to work on the docs as well (it's a new command). We need review from a core dev, especially on the interface. The main concern I have is about the conflict between backward-compatibility and readability:

```python
def post_mortem(traceback=None):
    # We allow(actually prefer) the input traceback to be an Exception.
```

How should we deal with that? Changing the argument name `traceback` would break some backward compatibility, enforcing the argument to be an `Exception` would break a lot.

This is a more serious concern for public interfaces, but also on private methods, for example, `Pdb.interaction` takes a `tb` argument which can be an `Exception` now. This is not good for readability.

Could @iritkatriel or @brandtbucher share some thoughts on this? Thanks!

Carreau commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,52 @@ can be overridden by the local file.
+
+         out()
+
+   calling ``pdb.pm()`` will allow to move between exceptions:
```
```suggestion
   calling ``pdb.pm()`` will allow to move between exceptions::
```

gaogaotiantian commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,52 @@ can be overridden by the local file.
+
+   List or jump between chained exceptions.
+
+   If ``Pdb.post_mortem()`` is called with a chained exception, it will now allow
```
Maybe mention `pm()` as well? It's probably a more common entry for this command.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -416,6 +418,17 @@ def preloop(self):
 
     def interaction(self, frame, traceback):
```
I think we should rename the argument `traceback` to something like `tb_or_exc` - it's internal so should not impact users.

gaogaotiantian commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,52 @@ can be overridden by the local file.
 
    Print the return value for the last return of the current function.
 
+.. pdbcommand:: exceptions [arg]
```
To keep the consistency, we should probably write `exceptions [excnumber]` instead of `exception [arg]`. That's now `clear`/`disable` documents.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1073,6 +1086,24 @@ def _select_frame(self, number):
         self.lineno = None
 
+    def do_exceptions(self, arg):
+        if not arg or arg == "list":
```
The `list` arg is not documented. Consider the fact that `break` command does not have a `list` option, maybe we can simply remove it.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -1912,11 +1948,7 @@ def post_mortem(t=None):
 
 def pm():
     """Enter post-mortem debugging of the traceback found in sys.last_traceback."""
```
The comment here should also be changed to `sys.last_exc` :)

gaogaotiantian reviewed (changes requested):
I think the overall direction is great and we should have this feature in pdb. However, we do need opinions from core devs, especially for the backward compatibility part. I'll wait for a couple of days and reach out to core devs if no one volunteers :)

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,27 @@ def preloop(self):
+            current = exception
+            while current := current.__context__:
+                self._chained_exception_index += 1
+                self._chained_exceptions.insert(0, current)
```
What about ``current.__cause__``?

Carreau replied:
My understanding is that `__context__` is always set, and `__cause__ == __context__` if `raise Err(..) from e`, so I _think_ we only need to look at `__context__`. 

```python
In [5]: try:
    ...:   try:
    ...:     raise ValueError('A')
    ...:   except Exception as e:
    ...:     raise TypeError('B') from e
    ...: except Exception as f:
    ...:     print(f.__context__, f.__cause__)
A A

In [6]: try:
    ...:   try:
    ...:     raise ValueError('A')
    ...:   except Exception as e:
    ...:     raise TypeError('B')
    ...: except Exception as f:
    ...:     print(f.__context__, f.__cause__)
A None
```

I originally was looking at both, but removed as `__context__` appeared to be enough. 
Unless we do want to have a visual distinction between when there is `__cause__` or not. 

Unless there is a case where we can have `__cause__` but not `__context__`, or where both are different, but I could not think of one. Though i'm far from an expert so if you think it's necessary I can check both.

iritkatriel replied:
The cause is whatever was in the "raise .. from". See below.

```
% cat xx.py 

import traceback

try:
    try:
        raise TypeError("bad type")
    except:
        raise ValueError("bad value") from OSError("something happened")
except Exception as e:
    exc = e

print("------------------- exc:")
traceback.print_exception(exc)
print("------------------- context:")
traceback.print_exception(exc.__context__)
print("------------------- cause:")
traceback.print_exception(exc.__cause__)
```


```
% ./python.exe xx.py
------------------- exc:
OSError: something happened

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/iritkatriel/src/cpython-1/xx.py", line 8, in <module>
    raise ValueError("bad value") from OSError("something happened")
ValueError: bad value
------------------- context:
Traceback (most recent call last):
  File "/Users/iritkatriel/src/cpython-1/xx.py", line 6, in <module>
    raise TypeError("bad type")
TypeError: bad type
------------------- cause:
OSError: something happened

```

Carreau replied:
Ah ! Thanks ! This is helpful. 

I've reworked the code to work with both `__cause__` and `__context__`.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,35 @@ def preloop(self):
+            open_list = [exception]
+            while open_list:
+                current = open_list.pop(0)
+                for ex_type in ["__cause__", "__context__"]:
```
In the standard traceback you get the context only if there is no cause. I think it makes sense here also to include the context only when there is no cause. See https://github.com/python/cpython/blob/fb08b7905eaf67902a89eeec0f8c1aa5ee0c7ea5/Lib/traceback.py#L740 .

Carreau replied:
Sure, changed to get into `__cause__` if present, otherwise `__context__` assuming `__supress_context__ == False`. Thus this should now work with `raise from None` (test added).

I tend to think that when using `pm()` you sort of really want to get to the root cause  of things, and it might be useful to get not have access to hidden context or cause. But I guess that can be added later if we wish.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,34 @@ def preloop(self):
+            while current:
+                self._chained_exception_index += 1
+                self._chained_exceptions.insert(0, current)
+                if current.__cause__:
```
```suggestion
                if current.__cause__ is not None:
```

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,34 @@ def preloop(self):
+                self._chained_exceptions.insert(0, current)
+                if current.__cause__:
+                    current = current.__cause__
+                elif current.__context__ and not current.__suppress_context__:
```
```suggestion
                elif current.__context__ is not None and not current.__suppress_context__:
```

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1073,6 +1093,24 @@ def _select_frame(self, number):
         self.lineno = None
 
+    def do_exceptions(self, arg):
+        if not arg:
```
what is the type of arg?

Carreau replied:
For archival purpose, see other comment, `arg` is a str, that represent the number of the current exception. Marking as resolved.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,34 @@ def preloop(self):
+            current = exception
+            while current:
+                self._chained_exception_index += 1
+                self._chained_exceptions.insert(0, current)
```
It's more efficient to append to a list and reverse the list in the end.

Carreau replied:
Done.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,34 @@ def preloop(self):
+            self._chained_exception_index = -1
+            current = exception
+            while current:
+                self._chained_exception_index += 1
```
Why do we need to maintain `_chained_exception_index` in this loop?

Carreau replied:
You are right, we can just use the length after populating the list.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,32 @@ def preloop(self):
+                    current = current.__context__
+                else:
+                    current = None
+        self._chained_exceptions = _exceptions[::-1]
```
```suggestion
        self._chained_exceptions = tuple(reversed(_exceptions))
```

Carreau replied:
I'm curious, why `tuple(reversed(...))` and not `[::-1]`. tuple to make immutable is better, but I though `[::-1]` was more efficient on list.

Carreau replied:
actually scratch that, wouldn't

```
_exceptions.reverse()
tuple(_exceptions)
```

be even faster ?

edit:and use less memory. Here I believe we _can_ use it as we are the only user of _exceptions_. Though style wise I don't like this I find it less readable.

iritkatriel replied:
My motivation here was readability rather than efficiency. It's all linear anyway.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,32 @@ def preloop(self):
+
+        _exceptions = []
+        if isinstance(tb_or_exc, BaseException):
+            tb_or_exc, exception = tb_or_exc.__traceback__, tb_or_exc
```
I think it's easier to follow if we keep the name of the traceback variable as is, and do this:

```suggestion
            traceback, exception = tb_or_exc.__traceback__, tb_or_exc
```

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,32 @@ def preloop(self):
+                elif current.__context__ is not None and not current.__suppress_context__:
+                    current = current.__context__
+                else:
+                    current = None
```
```suggestion
                    current = None
        else:
            traceback = tb_or_exc
```

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,16 +416,32 @@ def preloop(self):
             else:
                 Pdb._previous_sigint_handler = None
-        if self.setup(frame, traceback):
+        if self.setup(frame, tb_or_exc):
```
```suggestion
        if self.setup(frame, traceback):
```

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1073,6 +1091,34 @@ def _select_frame(self, number):
         self.print_stack_entry(self.stack[self.curindex])
         self.lineno = None
 
+    def do_exceptions(self, arg: str):
```
This module is not type annotated, so we shouldn't add this.
My question about the type was: what is "not arg" checking? For None? 0? Empty string?

Carreau replied:
Ah, empty strings, the `def fun(self, arg):` `if not arg` is used a couple of time in the module, but I can add a comment.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1890,11 +1936,16 @@ def set_trace(*, header=None):
     default is to be used).
+
+    If `t` is an Exception and is a chained exception (i.e it has a __context__),
+    pdb will be able to move to the sub-exception when reaching the bottom
```
```suggestion
    pdb will be able to move to other exceptions in the chain when reaching the bottom
```

(sub-exception sounds like it belongs to exception groups, and you're not dealing with those here).

iritkatriel replied:
That said, I don't think it's clear from the context what "when reaching the bottom" means.

Carreau replied:
Sorry that's a leftover when up/down were moving to previous/next exception in the chain when reaching the bottom/top of the stack, but we removed in in favor of do_exception. rephrased.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1073,6 +1093,35 @@ def _select_frame(self, number):
+        If given an integer as argument, switch to the exception at that index.
+        """
+        if not arg:
+            nwid = len(str(len(self._chained_exceptions) - 1))
```
I think we should just use 3 and somehow limit the number of exceptions in a chain you browse to 999. Truncate (with some msg) if there are more or something like that.

Which reminds me that we also need to handle cycles in the chain. Look at the traceback module to see how it's done there.

Carreau replied:
Ok, hardcoded to 3 digits (999) set on a private class variable of Pdb just in case.

At startup Pm stops after 999 exception and print a message that only the first 999 will be browsable with `exceptions`. I was not sure it was worth to make that lazy until called for in `do_exceptions`.

While I was also dealing with cycles, I added support for exceptions groups as well and added test with various exceptions, groups, cause, context...

iritkatriel commented on Lib/pdb.py:
```diff
@@ -422,15 +424,28 @@ def interaction(self, frame, tb_or_exc):
             traceback, exception = tb_or_exc.__traceback__, tb_or_exc
-            current = exception
-            while current:
+            currents = [exception]
```
I don't follow - what it ``currents``, and what is ``_exceptions``, and why do we need both?  Are you trying to allow navigating into exception groups? I think that would require some more design.

Carreau replied:
Sorry I might have taken your following comment too literally:

> Which reminds me that we also need to handle cycles in the chain. Look at the traceback module to see how it's done there.

In the tb module you use a `queue` and a set of `_seen` traceback object ids. I assumed that what you wanted me to do, but the queue there can only old multiple elements because it handles exceptions group I belive. So I assumed you pointed me there for me to also see the exception group.

I've remove the handling of exception groups, and reverted to the previous logic, simply check if the cause or context is in the exceptions we've already visited.

Unless I still misunderstand what you asked.

iritkatriel replied:
Sorry I wasn't clear. This is what I meant. This seems fine from the exceptions POV, but I'd like @gaogaotiantian to think about implications for PDB.   In particular, we are now saving a list of exceptions, which can create references to practically everything in the program through tracebacks (which reference frames). So what are the implications re memory leaks?

iritkatriel replied:
One option is to store a traceback.TracebackException object instead of the exception itself. These are objects that don't hold references to the whole world.

Carreau replied:
I _guess_ we can use a weaklist, as the result is anyway derived from `sys.last_exc`, there is at least this that keeps everything around and it is unlikely to disapear untl we exit the pm ?

iritkatriel replied:
I don't think you can create weak refs to exceptions.

Carreau replied:
or as the in end of `def interaction`, put the variable back to their initial state of empty tuple/0

gaogaotiantian replied:
During the PDB interaction, we are keeping all the frames anyway so I don't think having exceptions stored introduces any new issue. We should of course release the reference to the exceptions as long as we are out of PDB.

However, I do have some questions on this - why can't the release code be put in `forget()`? `setup()` calls `forget()` so we should extract the exception after `setup()`, other than that I did not see the calls to `forget()` during interaction, maybe I'm missing something?

Also, the code in `interaction` to get exceptions` is getting large here and I'd really like it to be another function.

It seems to me like `setup()` is doing very similar things. Should we have a new method like `setup_exceptions()` and call it in `setup()` if an exception is given? That's said, change the signature of `setup(self, f, tb)` to `setup(self, f, tb, exc=None)`, and do something like

```python
if exc is not None:
    self.setup_exceptions(exc)
```

in `setup()`.

I think this will make the `interaction()` function much clearer. Also we can do everything after the `forget()` in `setup()`. What do you think @iritkatriel ?

Carreau replied:
> However, I do have some questions on this - why can't the release code be put in `forget()`? `setup()` calls `forget()` so we should extract the exception after `setup()`, other than that I did not see the calls to `forget()` during interaction, maybe I'm missing something?

Every time we `(pdb) exceptions <number>` to a new exception we call `setup(new_ex)`. If we put the cleanup code in `forget` then as soon as we jump to a new exception we'll loose all our setup. It is also called in `bp_commands` as well but I'm not sure the effect of that.

I'm happy to pull setup/interactions into two functions, but I'm slightly worried that this would have subtle backward compatibility with upstream. 

> Also we can do everything after the forget() in setup()

Again as is it will currently break, as setup is called when we jump exceptions, and if we split into say setup/setup_one_exception forget/forget_all_exception we still risk breaking if any subclass call setup/forget in the wrong place – I can tell that at least 2 Pdb derived debugger don't call self.setup()/self.forget()

Carreau replied:
Adding onto this immediately, it looks like most setup/forget are in pairs, if we want to do a refactor that I would _prefer_ to do as a separate PR, I would change the setup/forget pair into a context manager.

I also pushed a commit that remove references to the exceptions after the command loop in interaction, and put a comment as to why it's not in forget.

gaogaotiantian replied:
My bad, `setup()` is to setup the frames and `exceptions` needs to setup a new set of frames when switching. Never mind.

Can we still pack the code setting up the exceptions to a separate function and call it from `interaction`? The cleaning part is compact and easy to follow at this point, but the setup piece is a bit large.

I actually have a separate question for @iritkatriel - we expect this feature mainly used with `pm` right? What's the most likely usage from the users? I would guess it's executing the code with CLI - `python -m pdb the_code.py`. However, the current implementation won't work on it because the `main()` function is not changed for the new feature. Should we change the `main()` function?

iritkatriel replied:
> Should we change the `main()` function?

I don't see why not.

gaogaotiantian replied:
Then let's change the main function as well, so that `python -m pdb raise_exception.py` could enter pm with exception list. Another minor issue - when there's no exception, `exceptions` command does nothing. I think we should tell the user that there's no exceptions or something like that - so the users know that they are not supposed to use the command. I know `break` does the same thing - I don't like that either but let's deal with it later. We should do something like `display` - let the users know what happened.

Carreau replied:
- Message added in do_exceptions if no chained exception found. 
- Main function modified to pass exception to post_mortem by default.
- Extracted the logic to create `self._chained_exceptions` into both a utility function and a context manager to make sure references are properly released.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -422,15 +424,28 @@ def interaction(self, frame, tb_or_exc):
+            while currents:
+                current = currents.pop()
+                if current in _exceptions:
+                    continue
```
If we put a limit of 999, maybe we shouldn't do anything special for cycles after all, just let them reach the limit?

Carreau replied:
Well if we have just A->B->C->A, it's still better to show "only" 3 exceptions than 999 ? 

I'm guessing the 999 limit might be more relevant for exception groups in async framework where you may have 1000+ tasks failing ?

Carreau commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,54 @@ can be overridden by the local file.
+   When using ``pdb.pm()``  or ``Pdb.post_mortem(...)`` with a chained exception
+   instead of a traceback, it will now allow the user to move between the
+   chained exceptions using ``exceptions`` command to list exceptions, and
+   ``exception <number>`` to switch to that exception.
```
Thoughts on adding something like this ? 
```suggestion
   ``exception <number>`` to switch to that exception. In the case of ``ExceptionGroup``s this will also allow to inspect sub exceptions.
```

iritkatriel replied:
I don't think this feature, as is, should allow traversing into exception groups. The arg implies a linear order of exceptions, and if you have an exception group in an exception chain, and each exception in the group can also be in a chain, it's not obvious how to flatten all this into a linear order.

Carreau replied:
Ok, removed.

Carreau committed:
Move the exception logic to its own function.

Also move the release of the list of exception to a context manager
for security

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -414,23 +419,84 @@ def preloop(self):
-        self._cmdloop()
-        self.forget()
+
+
```
We should keep the single line between methods.

Carreau replied:
Done.

gaogaotiantian commented on Lib/pdb.py:
```diff
@@ -206,6 +207,10 @@ def namespace(self):
+    # the max number of chained exceptions + exception groups we accept to navigate.
+    _max_chained_exception_depth = 999
+    _chained_exceptions = tuple()
+    _chained_exception_index = 0
```
Why do we put these in the class? Is there any case `_chained_exceptions` and `_chained_exception_index` would be used without proper initialization in the instances? `previous_sigint_handler` is used as `Pdb._previous_sigint_handler` and it's probably set for all `Pdb` instances, but the exceptions are per instance right? I think we should init it as instance members in `__init__`.

As for `_max_chained_exception_depth`, if that's an arbitrary constant without a documented way to change, we can consider that as a constant and use `MAX_CHAINED_EXCEPTION_DEPTH`?

Carreau replied:
I rarely see multiple Pdb instances used so I don't have any particular preferences. 

Though I don't quite like global variables, and I have the impression that attaching constant to the only class that use them make things a tiny bit more self contained and discoverable.

I still did both of the changes you suggested though.

gaogaotiantian replied:
I obviously misrepresented myself. I meant use `MAX_CHAINED_EXCEPTION_DEPTH` as a class member instead of `_max_chained_exception_depth`. Just change the variable name to indicate that this is a constant instead of a private variable. I still want to put it in `class Pdb` because that's the only scope where it would be used.

Carreau replied:
Ah, sure, I moved it back as a class variable.

gaogaotiantian reviewed (commented):
Overall I think this is good, and I know you have been working on this for a long time. I do have a couple of small concerns. Maybe you have thoughts that I did not consider.

Carreau committed:
Make variable instance variable.

And max chain depth a global one

AlexWaygood committed:
Merge branch 'main' into gh-106670

Carreau committed:
Move back MAX_CHAINED_EXCEPTION_DEPTH to class variable

iritkatriel commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,54 @@ can be overridden by the local file.
+   List or jump between chained exceptions.
+
+   When using ``pdb.pm()``  or ``Pdb.post_mortem(...)`` with a chained exception
+   instead of a traceback, it will now allow the user to move between the
```
```suggestion
   instead of a traceback, it allows the user to move between the
```

iritkatriel commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,54 @@ can be overridden by the local file.
+    > example.py(10)middle()
+    -> return inner(0)
+
+
```
We need a versionadded entry here to indicate that this is new in 3.13.

Carreau replied:
Added at the top just after `.. pdbcommand::` as all of this is new.

iritkatriel replied:
It typically goes in the end. You write the docs to describe the current situation and then add versionadded and versionchanged directives at the end. Look at other docs for examples.

Carreau replied:
ok. Should be done.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -205,10 +206,16 @@ def namespace(self):
 
+
+class Pdb(bdb.Bdb, cmd.Cmd):
+    # the max number of chained exceptions + exception groups we accept to navigate.
```
is this comment in the right place?

Carreau replied:
moved and rephrased.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -414,23 +424,82 @@ def preloop(self):
-            # a command like "continue")
+        _chained_exceptions, tb = self._get_tb_and_exceptions(tb_or_exc)
+        with self._hold_exceptions(_chained_exceptions):
+            if Pdb._previous_sigint_handler:
```
Does this need to be inside the _hold_exceptions() block?

Carreau replied:
Does not need to, I reduced the spam of the context manager.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1073,6 +1142,44 @@ def _select_frame(self, number):
+            self.message(
+                "Did not find chained exceptions. To move between"
+                " exceptions, pdb/post_mortem must be given an exception"
+                " object instead of a traceback."
```
```suggestion
                " object rather than a traceback."
```

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1890,11 +1997,16 @@ def set_trace(*, header=None):
 
 def post_mortem(t=None):
-    """Enter post-mortem debugging of the given *traceback* object.
+    """Enter post-mortem debugging of the given *traceback*, or *exception*
```
should we rename the parameter `t`?

Carreau replied:
That would be backward incompatible ... I'm always for renaming, if we do that I would even make it positional only:

```suggestion
def post_mortem(tb_or_exc=None, /):
```

Carreau replied:
renamed and made posonly.

iritkatriel replied:
> That would be backward incompatible ...

Good point. Leave it then.

iritkatriel replied:
> renamed and made posonly.

We can't do that. We would need to add the new option, deprecate the old, and support both for now. Not worth it.

Carreau replied:
Removed from the commit history.

iritkatriel commented on Lib/pdb.py:
```diff
@@ -1890,11 +1997,16 @@ def set_trace(*, header=None):
+
+    If `t` is an Exception and is a chained exception (i.e it has a __context__,
+    or a __cause__), pdb will be able to list and move to other exceptions in
+    the chain using the `exceptions` command
```
I think we just need a hint here, so tried to make it shorter (anyone interested in more details can check the docs).

```suggestion
    If `t` is an exception object, the `exceptions` command makes it possible to
    list and inspect its chained exceptions (if any).
```

iritkatriel commented on Misc/NEWS.d/next/Library/2023-07-12-10-59-08.gh-issue-106670.goQ2Sy.rst:
```diff
@@ -0,0 +1,3 @@
+Add the new ``exceptions`` command the Pdb debugger to move between chained exceptions when using post mortem debugging.
+
+
```
```suggestion
Add the new ``exceptions`` command to the Pdb debugger. It makes it possible to move between chained exceptions when using post mortem debugging.
```

iritkatriel commented on Doc/library/pdb.rst:
```diff
@@ -639,6 +639,54 @@ can be overridden by the local file.
 
    Print the return value for the last return of the current function.
 
+.. pdbcommand:: exceptions [excnumber]
```
We need to add an entry in Doc/whatsnew/3.13.rst  since this is a new feature.

Carreau replied:
Isn't that what `Misc/NEWS.d/next/Library/2023-07-12-10-59-08.gh-issue-106670.goQ2Sy.rst` is for ?

iritkatriel replied:
Not really. News is for the release manager. What'snew is for users.

Carreau replied:
Ok, I though this was auto-merged into the whatsnew. Added.

Carreau committed:
Apply suggestions from code review

Co-authored-by: Irit Katriel <1055913+iritkatriel@users.noreply.github.com>

Carreau committed:
Adress some of the reviews

Carreau committed:
add versionadded

Carreau committed:
Fix doc

Carreau commented on Lib/pdb.py:
```diff
@@ -1889,34 +1996,34 @@ def set_trace(*, header=None):
         exc = sys.exception()
         if exc is not None:
-            t = exc.__traceback__
+            tb_or_exc = exc.__traceback__
```
I _guess_ we do not need this anymore, and below we need to update the error message ?

Carreau commented on Lib/pdb.py:
```diff
@@ -1889,34 +1996,34 @@ def set_trace(*, header=None):
+    p.interaction(None, tb_or_exc)
 
 def pm():
     """Enter post-mortem debugging of the traceback found in sys.last_traceback."""
```
```suggestion
    """Enter post-mortem debugging of the traceback found in sys.last_exc."""
```

Carreau replied:
fixed that in one of the latests commits.

Carreau committed:
add whatsnew

iritkatriel commented on Doc/whatsnew/3.13.rst:
```diff
@@ -189,6 +189,14 @@ typing
+pdb
+---
+
+* Add ability to move between chained exceptions during post mortem debugging in :func:`pm()` using
```
There's a warning here about the func:pm reference. I don't think these references work in what's new. Maybe @CAM-Gerlach can advise.

AlexWaygood replied:
These references should work in "what's new".

You don't need the brackets in the `:func:` markup; they're added by sphinx when the markup is rendered. You also need to tell sphinx the module where the function is from, or sphinx can't add a link to it.

```suggestion
* Add ability to move between chained exceptions during post mortem debugging in :func:`~pdb.pm` using
```

Carreau replied:
Thanks !

AlexWaygood replied:
This docs error is now back after the latest rebase/squash

iritkatriel reviewed (approved):
LGTM, just need to resolve the warning in the news file.

Carreau committed:
Update Doc/whatsnew/3.13.rst

Co-authored-by: Alex Waygood <Alex.Waygood@Gmail.com>

iritkatriel commented on Doc/whatsnew/3.13.rst:
```diff
@@ -189,6 +189,14 @@ typing
+
+* Add ability to move between chained exceptions during post mortem debugging in :func:`~pdb.pm` using
+  the new ``exceptions [exc_number]`` command for Pdb. (Contributed by Matthias
+  Bussonnier in :gh:`106676`.)
```
Sorry one last nit. The modules here are sorted alphabetically, so this needs to be between path lib and sqlite.

Carreau replied:
No problem, done and rebased/squashed a bit for a cleaner history.

AlexWaygood replied:
> No problem, done and rebased/squashed a bit for a cleaner history.

Please don't use rebases or squashes on CPython PRs, when you can help it :) It interacts badly with the GitHub UI, making it hard for reviewers to see what changed between commits. Everything in CPython is squash-merged anyway, so we really don't care about messy commit history within PRs.

iritkatriel replied:
Thanks. 

In the future don't squash. We squash on merge anyway, so the commit history doesn't end up in the commit log. But it's useful to have it on the PR, and it's easier to review just the last commit and see what you've changed.

Carreau replied:
Ok, I can also un-squash as well I have the reflog.

AlexWaygood commented on Doc/whatsnew/3.13.rst:
```diff
@@ -158,6 +158,13 @@ pathlib
+pdb
+---
+
+* Add ability to move between chained exceptions during post mortem debugging in :func:`pm()` using
```
This will fix the failing docs build again:

```suggestion
* Add ability to move between chained exceptions during post mortem debugging in :func:`~pdb.pm` using
```

Carreau replied:
Damn, I screwd my rebase. Sorry, will redo.

AlexWaygood commented on Doc/whatsnew/3.13.rst:
```diff
@@ -189,6 +196,7 @@ typing
   check whether a class is a :class:`typing.Protocol`. (Contributed by Jelle Zijlstra in
   :gh:`104873`.)
 
+
```
A spurious newline has been added here:

```suggestion
```

Carreau committed:
move paragraph
